### PR TITLE
Add GitHub action workflow to automatically build/publish release artifacts after tag is created

### PR DESCRIPTION
**Pull request by jhump**
I tested this configuration with a personal repo to verify it does what we want. This allows us to hand-curate the release notes in the GitHub UI, but still have `goreleaser` create the artifacts that are part of the release.

This also adds a `RELEASE.md` doc to describe the release process, and it's similar to connect-go (which will likely soon have a similar "release" action to publish the `protoc-gen-connect-go` binary as a release artifact) and connect-kotlin (which uses a "release" action to publish to Maven central).

### CODE REVIEW

**Comment by jhump** [comment]
Ping, @smaye81.

After #826 is merged, we should be ready for v1.0.0-rc4 (hopefully final release candidate). And I'd like to try out this change with that next release.

File: RELEASE.md
Comment (akshayjshah) [inline RELEASE.md:17]: Can we specifically mention that pre-1.0 releases are _not_ automatically prereleases? Across our OSS, we settled on treating `v0.9.5` as a regular release. I assume that also applies to RCs for the 1.0, since we _do_ want to feature those releases on the repository homepage?

File: .github/workflows/ci.yaml
Comment (akshayjshah) [inline .github/workflows/ci.yaml:15]: Can we leave a comment reminding maintainers to also keep `release.yaml` updated when changing Go versions?

**Review by akshayjshah — APPROVED** [review]

File: RELEASE.md
Comment (jhump) [reply RELEASE.md]: I didn't realize that RCs should not be marked as pre-releases. For Go libraries anyway, using `go get -u` won't get release candidates, so it seemed okay to align the GitHub release state the same way, by marking them as pre-releases.

I can wordsmith a little on this.

File: .github/workflows/ci.yaml
Comment (jhump) [reply .github/workflows/ci.yaml:15]: Great idea

**Comment by jhump** [comment]
@akshayjshah, I was just looking at this and realized I totally forgot to describe the process to update the embedded version number via PR. So I just pushed a commit with more words, mostly copied from similar language for connect-go releases.

File: RELEASE.md
Comment (jhump) [reply RELEASE.md]: I added a single statement about pre-v1.0.0 releases. But would like to confirm: do we want release candidates to _not_ be release candidates, too?

Comment (akshayjshah) [reply RELEASE.md]: As far as I can tell, the only effect of this checkbox for Go libraries and binaries is controlling whether or not they're featured on the repository homepage. It seems that we want to advertise the existence of v1.0-rc4, right? We might feel differently about RCs for v2, if we ever get there - we'd want most users to continue using 1.x releases instead of blindly updating to something unstable.

**Review by akshayjshah — APPROVED** [review]

**Pull request merged**
